tsdb/shard.go

```diff
@@ -1302,6 +1306,143 @@ func (a Shards) FieldKeysByPredicate(expr influxql.Expr) (map[string][]string, e
 	return all, nil
 }
 
+type measurementEvaluator struct {
+	measurementNodeHeads map[*influxql.BinaryExpr][]string
+	invalidHeads         map[*influxql.BinaryExpr]struct{}
+	onlyAnds             bool
+	head                 *influxql.BinaryExpr
+	measurementKey       string
+}
+
+// Visit is called on every node to evaluate the tree for a possible measurement
+// optimization. A measurement optimization is possible if the tree contains a
+// single group of one or many measurements (in the form of _measurement =
+// measName, equality operator only) grouped together by OR operators, with the
+// subtree containing the OR'd measurements accessible from root of the tree
+// either directly (tree contains nothing but OR'd measurements) or by
+// traversing AND binary expression nodes.
+func (v measurementEvaluator) Visit(node influxql.Node) influxql.Visitor {
```

Comment:
Would it make the logic better to have two visitors? You could have one visitor responsible for finding a potential head ( the LHS or RHS of an AND node that was not itself an AND node) and a second visitor responsible for finding if a given head is valid (composed of only measurement EQ conditions) or fatal (some measurement conditions and some non-measurements).

Separating the complicated algorithm into two simple algorithms would probably make the whole thing more clear when we read this code in the future.

Reply:
I split it up and I think it did make it easier to understand. Check it out and see what you think!